gh-80192: Use windows api if ssl cert verification fails

tobil4sk opened the pull request:
This avoids rejecting certificates that should be accepted, see #80192.

Here is the relevant openssl documentation: https://docs.openssl.org/master/man3/SSL_CTX_set_verify/.

Also, here are the windows api docs:
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certverifycertificatechainpolicy
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_policy_status for error status.

I've been testing manually by deleting `DigiCert Global Root G2` from `Third Party Root Certificates` in the user certificate store (this is safe, the certificate will be added back automatically), and running the following code:

```python
import http.client
conn = http.client.HTTPSConnection("www.verisign.com")
conn.request("GET", "/")
conn.close()
```

The main branch errors with:
```
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1025)
```
Whereas with this patch, the certificate is verified and the connection is established.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-80192 -->
* Issue: gh-80192
<!-- /gh-issue-number -->

tobil4sk committed:
Use windows api if ssl cert verification fails

This avoids rejecting certificates that should be accepted, see #80192

tobil4sk committed:
Use const where appropriate

ZeroIntensity commented on Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst:
```diff
@@ -0,0 +1 @@
+Fixed valid ssl certificates being rejected.
```
Use a reference:
```suggestion
Fixed valid :mod:`ssl` certificates being rejected.
```

tobil4sk replied:
Reference added

asvetlov replied:
I would see a broader text, e.g. the PR verifies certifications rejected by OpenSSL by additional calls to Windows specific API functions.
The exact proposed text is definitely not fine, but you have got my idea.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2996,6 +2996,119 @@ static PyType_Spec PySSLSocket_spec = {
+        return NULL;
+    }
+
+    cert_bytes = PyMem_RawMalloc(cert_bytes_length);
```
Is it possible to use pymalloc (`PyMem_Malloc`) here? Typically, we only need `RawMalloc` if we plan on using it inside `Py_BEGIN_ALLOW_THREADS` areas.

tobil4sk replied:
Using PyMem_Malloc I got the following error in my sample program:
`Fatal Python error: _PyMem_DebugMalloc: Python memory allocator called without holding the GIL`

ZeroIntensity replied:
I figured. It might be worth adding a comment mentioning that the GIL isn't held (so no Python APIs can get called).

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2996,6 +2996,119 @@ static PyType_Spec PySSLSocket_spec = {
+        return preverify_ok;
+    }
+
+    HCERTSTORE store = CertOpenStore(CERT_STORE_PROV_MEMORY, 0, 0, CERT_STORE_DEFER_CLOSE_UNTIL_LAST_FREE_FLAG, NULL);
```
There's a lot of lines in this function that are too long (the guideline is 79 characters, per PEP-7). Could you break them up across multiple lines?

tobil4sk replied:
I've fixed most violations except line 3063 which has a long constant name and is a couple characters too long, does it make more sense there to enforce the rule or to leave it?

ZeroIntensity replied:
I think it's fine.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2996,6 +2996,119 @@ static PyType_Spec PySSLSocket_spec = {
+        break;
+    case CERT_E_WRONG_USAGE:
+    case CERT_E_CRITICAL:
+    case CERT_E_PURPOSE:
```
There's a chance we need to use `_Py_FALLTHROUGH` here, but it looks like build isn't complaining. If it does start failing feel free to use that.

ZeroIntensity reviewed (commented):
Yeah, this needs a test. FYI, `test_ssl` is failing.

tobil4sk committed:
Refactor error translation into separate function

tobil4sk committed:
Fix PEP 7 violations

tobil4sk committed:
Add reference to news entry

tobil4sk committed:
Fix error translation resulting in test fail

CERT_E_UNTRUSTEDROOT:
A certification chain processed correctly but terminated in a root
certificate that is not trusted by the trust provider.

X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY:
The issuer certificate could not be found: this occurs if the issuer
certificate of an untrusted certificate cannot be found.

- https://docs.openssl.org/master/man3/X509_STORE_CTX_get_error/#error-codes
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_policy_status#members

tobil4sk commented on Modules/_ssl.c:
```diff
@@ -2996,6 +2996,125 @@ static PyType_Spec PySSLSocket_spec = {
+        return X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT;
+    case CERT_E_EXPIRED:
+        return X509_V_ERR_CERT_HAS_EXPIRED;
+    case CERT_E_INVALID_POLICY:
+        return X509_V_ERR_INVALID_POLICY_EXTENSION;
```
This patch seems to satisfy the other failing check in `test_ssl`, however, I'm not sure whether this is correct:

```suggestion
    case CERT_E_INVALID_POLICY:
        return X509_V_ERR_INVALID_CA;
```

Here are the relevant docs:

CERT_E_INVALID_POLICY:
The certificate has an invalid policy.

X509_V_ERR_INVALID_CA:
A CA certificate is invalid. Either it is not a CA or its extensions are
not consistent with the supplied purpose.


- https://docs.openssl.org/master/man3/X509_STORE_CTX_get_error/#error-codes
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_policy_status#members

tobil4sk replied:
It looks like test_ssl isn't failing anymore actually, however, test_poplib is failing due to a different error

zooba commented on Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst:
```diff
@@ -0,0 +1 @@
+Fixed valid :mod:`ssl` certificates being rejected.
```
Agree with Andrew's comment - this needs to be significantly expanded, because this is a very impactful change.

We also would need a section in `Doc/whatsnew/3.14.rst` explaining the impact and options for workarounds.

We _might_ also need to use a deprecation period to turn this on by default, though part of me would prefer not to.
